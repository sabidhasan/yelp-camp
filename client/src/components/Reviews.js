import React from 'react'
import RatingBar from './RatingBar'
import { formatDate } from '../helpers/helpers'
import PropTypes from 'prop-types'
import ReactCSSTransitionGroup from 'react-addons-css-transition-group'

class Reviews extends React.Component {
  constructor(props) {
    super(props);
  }

  static contextTypes = {
    user: PropTypes.object,
  };

  render() {
    if (!this.props.comments || this.props.comments.length === 0) {
      return <h2 className='review__first'>Be the first to review this campground</h2>
    }

    const allComments = this.props.comments.map(val => {
      const showDeleteButton = this.context.user && this.context.user.uid === val.uid;
      return (
        <div className='review' key={val.id}>
          <span className='review__author'>
            <img src={val.photoURL} alt='' />
            <span className='review__author-name'>{val.displayName}</span>
          </span>
          <div className='review__rating rating'>
            <RatingBar rating={val.rating} />
            <span className='review__time'>{formatDate(new Date(val.time))}</span>
          </div>
          {showDeleteButton ?
            <a className='review__delete' onClick={() => this.props.deleteReview(val.id)}>X</a>
            : null
          }
          <span className='review__text'>{val.text}</span>
        </div>
      )
    })

    return (
      <ReactCSSTransitionGroup
          transitionName="review-delete"
          transitionEnterTimeout={300}
          transitionLeaveTimeout={300}
          className={'reviews-container'}
          component='div'
      >
      {allComments}
    </ReactCSSTransitionGroup>
  )
  }
}

export default Reviews
