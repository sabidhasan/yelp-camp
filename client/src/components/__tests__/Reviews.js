import React from 'react'
import toJson from 'enzyme-to-json'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import Reviews from '../Reviews'

configure({ adapter: new Adapter() });

describe('<Reviews />', () => {
  let component;
  let props;
  let deleteReview;

  beforeEach(() => {
    deleteReview = jest.fn();

    props = {
      comments: [
        {id: 1, photoURL: 'url', displayName: 'Name', rating: 3.3, time: 'October 1, 2018', deleteReview, text: 'Test', uid: '10'},
        {id: 2, photoURL: 'url', displayName: 'Name', rating: 3.3, time: 'October 1, 2018', deleteReview, text: 'Test', uid: '10'},
        {id: 3, photoURL: 'url', displayName: 'Name', rating: 3.3, time: 'October 1, 2018', deleteReview, text: 'Test', uid: '10'},
        {id: 4, photoURL: 'url', displayName: 'Name', rating: 3.3, time: 'October 1, 2018', deleteReview, text: 'Test', uid: '10'},
      ],
      deleteReview
    }
    component = shallow(<Reviews { ...props } />)
  })

  it('Component should be defined', () => {
    expect(component).toBeDefined();
  });

  it('If no comments, then `Be the first to review` message shown', () => {
    component.setProps({comments: undefined});
    expect(component.find('.Reviews__first')).toHaveLength(1)
  });

  it('Number of reviews shown is equal to number of comments', () => {
    expect(component.find('Review')).toHaveLength(4)
  })

  it('Reviews component snapshot matches what is expected', () => {
    expect(toJson(component)).toMatchSnapshot();
  })
})
