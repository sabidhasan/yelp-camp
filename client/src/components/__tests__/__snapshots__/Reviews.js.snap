// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Reviews /> Reviews component snapshot matches what is expected 1`] = `
<CSSTransitionGroup
  className="Reviews"
  component="ul"
  transitionAppear={false}
  transitionEnter={true}
  transitionEnterTimeout={300}
  transitionLeave={true}
  transitionLeaveTimeout={300}
  transitionName="Reviews-delete"
>
  <Review
    deleteReview={[Function]}
    displayName="Name"
    id={1}
    key="1"
    photoURL="url"
    rating={3.3}
    text="Test"
    time="October 1, 2018"
    uid="10"
  />
  <Review
    deleteReview={[Function]}
    displayName="Name"
    id={2}
    key="2"
    photoURL="url"
    rating={3.3}
    text="Test"
    time="October 1, 2018"
    uid="10"
  />
  <Review
    deleteReview={[Function]}
    displayName="Name"
    id={3}
    key="3"
    photoURL="url"
    rating={3.3}
    text="Test"
    time="October 1, 2018"
    uid="10"
  />
  <Review
    deleteReview={[Function]}
    displayName="Name"
    id={4}
    key="4"
    photoURL="url"
    rating={3.3}
    text="Test"
    time="October 1, 2018"
    uid="10"
  />
</CSSTransitionGroup>
`;
